File: scripts/saas_sync.py
"""
scripts/saas_sync.py

Módulo compartido: sincroniza leads GA al SaaS API (POST /api/sync).
Importar desde leads_atlanta_premium.py via:
    sys.path.append(str(Path(__file__).parent))
    from saas_sync import sync_ga_leads

Variables de entorno requeridas:
    SAAS_API_URL   — URL base de la app (ej: https://tu-app.vercel.app)
    SAAS_API_KEY   — Clave secreta configurada en el SaaS (.env SYNC_API_KEY)

Opcional:
    SAAS_SCRAPER_SOURCE — string identificador del scraper (default: multivenza-ga)
"""

import json
import os
import urllib.request
import urllib.error
from datetime import datetime
from typing import Any, Dict, List, Optional

# Carga .env si python-dotenv está instalado
try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass

SAAS_API_URL    = os.getenv("SAAS_API_URL", "").rstrip("/")
SAAS_API_KEY    = os.getenv("SAAS_API_KEY", "")
SCRAPER_SOURCE  = os.getenv("SAAS_SCRAPER_SOURCE", "multivenza-ga")

# Mapeo de niche_category (GA) → project_type del SaaS
_NICHE_MAP: Dict[str, str] = {
    "ROOFING":            "Roofing",
    "NEW_CONSTRUCTION":   "New Construction",
    "COMMERCIAL":         "CGC",
    "REMODELING":         "Remodel",
    "SOLAR":              "Remodel",
    "BASEMENT":           "Remodel",
    "BASEMENT_FINISHING": "Remodel",
    "ADDITION":           "New Construction",
    "HVAC":               "HVAC",
    "OTHER":              "Remodel",
}

# Nombres placeholder que NO deben enviarse como owner_name real
_FAKE_NAMES = frozenset({
    "", "N/A", "NONE", "UNKNOWN", "PENDING VERIFICATION",
    "GWINNETT CLIENT", "FULTON CLIENT", "ATLANTA CLIENT",
})


# ── Helpers ───────────────────────────────────────────────────────────────────

def _tier(tpv: float) -> str:
    """Diamond si TPV > $15k (regla del SaaS)."""
    if tpv > 15_000:
        return "diamond"
    if tpv > 5_000:
        return "premium"
    return "standard"


def _score(lead: Dict[str, Any]) -> int:
    """Score 0-100 derivado de los datos disponibles en leads GA."""
    s = 10
    tpv = float(lead.get("tpv") or lead.get("valuation") or 0)

    niche = (lead.get("niche_category") or "").upper()
    if "ROOF" in niche:
        s += 20
    elif "COMMERCIAL" in niche or "CGC" in niche:
        s += 15
    elif "NEW_CONSTRUCTION" in niche or "ADDITION" in niche:
        s += 10

    no_gc = not bool(
        (lead.get("contractorName") or "").strip() or
        (lead.get("contractorId")   or "").strip()
    )
    if no_gc:
        s += 40

    if tpv > 250_000:
        s += 15
    elif tpv > 50_000:
        s += 10
    elif tpv > 15_000:
        s += 5

    if lead.get("tier") == "HIGH-TICKET":
        s += 10

    return min(s, 100)


def _map_lead(lead: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """Convierte un lead GA procesado al shape del SaaS API."""
    permit_number = (lead.get("leadId") or "").strip()
    if not permit_number:
        return None  # sin clave de deduplicación → omitir

    tpv = float(lead.get("tpv") or lead.get("valuation") or 0)

    niche        = (lead.get("niche_category") or "OTHER").upper().replace(" ", "_")
    project_type = _NICHE_MAP.get(niche, "Remodel")

    # Tags
    raw_tags  = lead.get("tags", "")
    tags: List[str] = (
        [t.strip() for t in raw_tags.split(",") if t.strip()]
        if isinstance(raw_tags, str) else list(raw_tags or [])
    )
    if "GA" not in tags:
        tags.insert(0, "GA")

    # Fecha de permiso normalizada a YYYY-MM-DD
    permit_date = lead.get("permitDate") or lead.get("permit_date")
    if permit_date and "T" in str(permit_date):
        permit_date = str(permit_date)[:10]

    # Owner — descartar nombres falsos
    owner = (lead.get("ownerName") or "").strip()
    owner = owner if owner.upper() not in _FAKE_NAMES else None

    no_gc = not bool(
        (lead.get("contractorName") or "").strip() or
        (lead.get("contractorId")   or "").strip()
    )

    return {
        "city":                lead.get("city") or "",
        "zip_code":            lead.get("zip")  or None,
        "state":               "GA",
        "county":              lead.get("county") or None,
        "project_type":        project_type,
        "estimated_valuation": tpv,
        "tier":                _tier(tpv),
        "score":               _score(lead),
        "tags":                tags,
        "no_gc":               no_gc,
        "roof_age":            None,
        "roof_classification": None,
        "permit_status":       lead.get("status") or None,
        "market_note":         lead.get("market_note") or None,
        "exact_address":       lead.get("address") or None,
        "owner_name":          owner,
        "phone":               lead.get("phone") or None,
        "contractor_name":     lead.get("contractorName") or None,
        "permit_number":       permit_number,
        "permit_date":         permit_date or None,
        "government_source":   lead.get("source") or "Accela ACA / Tyler EnerGov (GA)",
        "processed_at":        lead.get("processedAt") or datetime.now().isoformat(),
    }


def _post_json(url: str, payload: dict, extra_headers: dict) -> Optional[dict]:
    data = json.dumps(payload, ensure_ascii=False).encode("utf-8")
    req  = urllib.request.Request(url, data=data, method="POST")
    req.add_header("Content-Type",   "application/json; charset=utf-8")
    req.add_header("Content-Length", str(len(data)))
    for k, v in extra_headers.items():
        req.add_header(k, v)
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            return json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
        print(f"[saas_sync] HTTP {e.code}: {body[:400]}")
        return None
    except Exception as exc:
        print(f"[saas_sync] Error de red: {exc}")
        return None


# ── Público ───────────────────────────────────────────────────────────────────

def sync_ga_leads(
    leads: List[Dict[str, Any]],
    batch_id: Optional[str] = None,
) -> Optional[dict]:
    """
    Sincroniza leads GA procesados al SaaS API.

    Args:
        leads:    Lista de dicts producidos por leads_atlanta_premium.py → process_permits()
        batch_id: Identificador del lote (default: GA-YYYY-MM-DD)

    Returns:
        Dict con {inserted, updated, skipped, errors} o None si hubo error de red.
    """
    if not SAAS_API_URL or not SAAS_API_KEY:
        print("[saas_sync] ⚠️  SAAS_API_URL o SAAS_API_KEY no configurados — sync omitido")
        print("            Agrega estas variables a tu .env para activar la sincronización")
        return None

    mapped = [m for lead in leads if (m := _map_lead(lead)) is not None]

    if not mapped:
        print("[saas_sync] No hay leads GA válidos para sincronizar (ninguno tiene leadId)")
        return None

    url     = f"{SAAS_API_URL}/api/sync"
    payload = {
        "source_state": "GA",
        "batch_id":     batch_id or f"GA-{datetime.now().strftime('%Y-%m-%d')}",
        "leads":        mapped,
    }

    print(f"\n[saas_sync] → Enviando {len(mapped)} leads GA a {url} ...")

    result = _post_json(url, payload, {
        "x-api-key":        SAAS_API_KEY,
        "x-scraper-source": SCRAPER_SOURCE,
    })

    if result:
        ins  = result.get("inserted", 0)
        upd  = result.get("updated",  0)
        skip = result.get("skipped",  0)
        errs = result.get("errors",   [])
        print(f"[saas_sync] ✅ GA sync: {ins} nuevos · {upd} actualizados · {skip} omitidos")
        if errs:
            print(f"[saas_sync] ⚠️  Errores parciales: {errs[:3]}")
    return result

File: scripts/lib/saas_sync.js
'use strict';

/**
 * scripts/lib/saas_sync.js
 *
 * Módulo compartido: sincroniza leads FL e IL al SaaS API (POST /api/sync).
 *
 * Variables de entorno requeridas:
 *   SAAS_API_URL   — URL base de la app (ej: https://tu-app.vercel.app)
 *   SAAS_API_KEY   — Clave secreta configurada como SYNC_API_KEY en el SaaS
 *
 * Opcional:
 *   SAAS_SCRAPER_SOURCE — string identificador del scraper
 *
 * Uso desde run_diamond_leads.js:
 *   const { syncFLLeads } = require('./lib/saas_sync');
 *   await syncFLLeads(leads, batchId);
 *
 * Uso desde extract_chicago.js:
 *   const { syncILLeads } = require('./lib/saas_sync');
 *   await syncILLeads(results, batchId);
 */

const https = require('https');
const http  = require('http');
const { URL } = require('url');

const SAAS_API_URL   = (process.env.SAAS_API_URL || '').replace(/\/$/, '');
const SAAS_API_KEY   = process.env.SAAS_API_KEY || '';
const SCRAPER_SOURCE = process.env.SAAS_SCRAPER_SOURCE || 'multivenza-leadhunter';

// ── Tier ──────────────────────────────────────────────────────────────────────

/** Diamond si TPV > $15k (regla del SaaS). */
function _tier(tpv) {
  if (tpv > 70_000) return 'diamante';
  if (tpv >= 30_000) return 'oro';
  return 'plata';
}

// ── Mappers ───────────────────────────────────────────────────────────────────

const _FL_CATEGORY_MAP = {
  roofing:      'Roofing',
  cgc:          'CGC',
  homeBuilders: 'New Construction',
  other:        'Remodel',
};

/**
 * Convierte un lead FL procesado (output de Processor.run) al shape del SaaS API.
 * @param {object} lead  — lead object de processor.js
 */
function _mapFLLead(lead, sourceState) {
  // totalProjectValue incluye el piso de $250k para ciudades PREMIUM — es el
  // valor real del contrato, que el SaaS debe mostrar como estimated_valuation
  const tpv = lead.projectValue?.totalProjectValue ?? lead.valuation ?? 0;

  return {
    city:                lead.city         || '',
    zip_code:            lead.zip          || null,
    state:               lead.state        || sourceState || (
      lead.city?.toUpperCase().includes('HOUSTON') ? 'TX' :
      lead.city?.toUpperCase().includes('PHOENIX') ? 'AZ' :
      lead.city?.toUpperCase().includes('ATLANTA') ? 'GA' : 'FL'
    ),
    county:              lead.county       || null,
    project_type:        _FL_CATEGORY_MAP[lead.category] || 'Remodel',
    estimated_valuation: tpv,
    tier:                lead.tier ? lead.tier.toLowerCase() : _tier(tpv),
    projected_profit:    lead.projectValue?.estNetProfit ?? (tpv > 70000 ? 500 : tpv >= 30000 ? 250 : 125),
    score:               lead.score        ?? 0,
    tags:                lead.tags         ?? [],
    no_gc:               lead.flags?.noGC  ?? false,
    roof_age:            lead.roofAnalysis?.age            ?? null,
    roof_classification: lead.roofAnalysis?.classification ?? null,
    permit_status:       lead.status       || null,
    market_note:         lead.urgency?.message             ?? null,
    // Dirección normalizada preferida sobre la raw
    exact_address:       lead.addressFormatted || lead.address || null,
    owner_name:          lead.ownerName    || null,
    phone:               null,   // se llenará vía Outscraper / BatchData en paso posterior
    contractor_name:     lead.contractorName || null,
    permit_number:       lead.permitNumber,
    permit_date:         lead.permitDate   || null,
    government_source:   lead.source       || null,
    processed_at:        lead.processedAt  || new Date().toISOString(),
  };
}

const _IL_TYPE_MAP = {
  'PORCH_CONSTRUCTION':   'Remodel',
  'BASEMENT_FINISHING':   'Remodel',
  'COMMERCIAL_BUILD-OUT': 'CGC',
  'ROOFING':              'Roofing',
  'COMMERCIAL_BUILDOUT':  'CGC',
};

// Nombres IL que son claramente placeholders
const _FAKE_IL_OWNERS = new Set([
  '', 'n/a', 'none', 'unknown', 'chicago client 0', 'chicago client 1',
  'chicago client 2', 'chicago client 3', 'chicago client 4',
]);

/**
 * Convierte un lead IL (output de extract_chicago.js) al shape del SaaS API.
 * @param {object} lead
 */
function _mapILLead(lead) {
  const valuation    = parseFloat(lead.Valuation || 0);
  const projectType  = _IL_TYPE_MAP[lead.Fast_Cash_Type] || 'Remodel';

  const rawOwner = (lead.Owner_Name || '').trim().toLowerCase();
  const owner    = _FAKE_IL_OWNERS.has(rawOwner) ? null : lead.Owner_Name;

  return {
    city:                lead.City      || '',
    zip_code:            lead.ZIP       || null,
    state:               'IL',
    county:              lead.County    || null,
    project_type:        projectType,
    estimated_valuation: valuation,
    tier:                _tier(valuation),
    score:               _ilScore(lead, valuation),
    tags:                ['IL', lead.Is_Chicago ? 'CHICAGO' : null].filter(Boolean),
    no_gc:               false,
    roof_age:            null,
    roof_classification: null,
    permit_status:       lead.Status    || null,
    market_note:         null,
    exact_address:       lead.Address   || null,
    owner_name:          owner,
    phone:               null,
    contractor_name:     null,
    permit_number:       lead.Permit_Number,
    permit_date:         lead.Fecha_Permiso || null,
    government_source:   'Chicago Data Portal (Cook County)',
    processed_at:        new Date().toISOString(),
  };
}

function _ilScore(lead, valuation) {
  let s = 10;
  const type = (lead.Fast_Cash_Type || '').toUpperCase();
  if (type.includes('COMMERCIAL'))  s += 15;
  if (type.includes('ROOFING'))     s += 20;
  if (valuation > 50_000) s += 15;
  else if (valuation > 15_000) s += 10;
  return Math.min(s, 100);
}

// ── HTTP helper (sin dependencias externas) ───────────────────────────────────

function _postJSON(url, body, headers) {
  return new Promise((resolve, reject) => {
    const parsed  = new URL(url);
    const payload = JSON.stringify(body);
    const lib     = parsed.protocol === 'https:' ? https : http;

    const options = {
      hostname: parsed.hostname,
      port:     parsed.port || (parsed.protocol === 'https:' ? 443 : 80),
      path:     parsed.pathname + (parsed.search || ''),
      method:   'POST',
      headers: {
        'Content-Type':   'application/json; charset=utf-8',
        'Content-Length': Buffer.byteLength(payload),
        ...headers,
      },
    };

    const req = lib.request(options, res => {
      let data = '';
      res.on('data', chunk => { data += chunk; });
      res.on('end', () => {
        try { resolve({ status: res.statusCode, body: JSON.parse(data) }); }
        catch { resolve({ status: res.statusCode, body: data }); }
      });
    });
    req.on('error', reject);
    req.write(payload);
    req.end();
  });
}

// ── Core sync ─────────────────────────────────────────────────────────────────

async function _sync(mappedLeads, mappedCompetitors, state, batchId) {
  if (!SAAS_API_URL || !SAAS_API_KEY) {
    console.warn('[saas_sync] ⚠️  SAAS_API_URL o SAAS_API_KEY no configurados — sync omitido');
    return null;
  }

  const leads = mappedLeads || [];
  const competitors = mappedCompetitors || [];

  if (!leads.length && !competitors.length) {
    console.warn(`[saas_sync] No hay data ${state} para sincronizar`);
    return null;
  }

  const url = `${SAAS_API_URL}/api/sync`;
  const CHUNK_SIZE = 500; // Chunk size to avoid 4.5MB Vercel limit
  let totalInserted = 0, totalUpdated = 0, totalSkipped = 0;
  
  // Calculate max iterations based on largest array
  const maxIters = Math.max(
    Math.ceil(leads.length / CHUNK_SIZE),
    Math.ceil(competitors.length / CHUNK_SIZE)
  ) || 1;

  console.log(`\n[saas_sync] → Iniciando envío de ${leads.length} leads y ${competitors.length} competidores (${state}) a ${url} en ${maxIters} lotes...`);

  for (let i = 0; i < maxIters; i++) {
    const chunkLeads = leads.slice(i * CHUNK_SIZE, (i + 1) * CHUNK_SIZE);
    const chunkCompetitors = competitors.slice(i * CHUNK_SIZE, (i + 1) * CHUNK_SIZE);
    
    if (!chunkLeads.length && !chunkCompetitors.length) continue;

    const payload = {
      source_state: state,
      batch_id:     batchId || `${state}-${new Date().toISOString().slice(0, 10)}`,
      leads:        chunkLeads,
      competitors:  chunkCompetitors,
    };

    try {
      const res = await _postJSON(url, payload, {
        'x-api-key':        SAAS_API_KEY,
        'x-scraper-source': SCRAPER_SOURCE,
      });

      if (res.status === 200) {
        const { inserted = 0, updated = 0, skipped = 0, errors = [] } = res.body;
        totalInserted += inserted;
        totalUpdated += updated;
        totalSkipped += skipped;
        console.log(`[saas_sync] Lote ${i+1}/${maxIters} ✅: ${inserted} insertados, ${updated} actualizados`);
        if (errors.length) {
          console.warn(`[saas_sync] ⚠️  Errores en lote: ${errors.slice(0, 2).join(', ')}`);
        }
      } else {
        console.error(`[saas_sync] HTTP ${res.status} en lote ${i+1}:`, JSON.stringify(res.body).slice(0, 200));
      }
    } catch (err) {
      console.error(`[saas_sync] Error de red en lote ${i+1}: ${err.message}`);
    }
  }

  console.log(`[saas_sync] 🎉 Sincronización completa: ${totalInserted} leads insertados, ${totalUpdated} actualizados.`);
  return { inserted: totalInserted, updated: totalUpdated, skipped: totalSkipped };
}

// ── Exports ───────────────────────────────────────────────────────────────────

/**
 * Sincroniza leads FL procesados (output de Processor.run) al SaaS API.
 * @param {object[]} leads   — array de leads de Processor.run()
 * @param {object[]} competitors — array de competidores 
 * @param {string}   batchId — identificador del lote (ej: "FL-2026-04-29")
 */
async function syncFLLeads(leads, competitors, batchId) {
  return _sync(leads.map(l => _mapFLLead(l, 'FL')), competitors, 'FL', batchId);
}

/**
 * Sincroniza leads IL (output de extract_chicago.js) al SaaS API.
 * @param {object[]} leads   — array de leads de extract_chicago.js
 * @param {string}   batchId — identificador del lote
 */
async function syncILLeads(leads, batchId) {
  // Filtrar leads sin permit_number válido
  const valid = leads.filter(l => l.Permit_Number);
  return _sync(valid.map(_mapILLead), [], 'IL', batchId);
}

/**
 * Sincroniza leads FL procesados a Supabase (alias de syncFLLeads).
 */
async function syncToSupabase(leads, competitors, batchId) {
  return syncFLLeads(leads, competitors, batchId);
}

module.exports = { syncFLLeads, syncILLeads, syncToSupabase };
